src/main/kotlin/io/kotlintest/TestRunner.kt

```diff
@@ -0,0 +1,62 @@
+package io.kotlintest
+
+import org.junit.platform.engine.EngineExecutionListener
+import org.junit.platform.engine.TestExecutionResult
+import org.junit.runners.model.TestTimedOutException
+import java.util.concurrent.Executors
+
+class TestRunner(val listener: EngineExecutionListener) {
+
+  private fun <CONTEXT> createInterceptorChain(
+      interceptors: Iterable<(CONTEXT, () -> Unit) -> Unit>,
+      initialInterceptor: (CONTEXT, () -> Unit) -> Unit): (CONTEXT, () -> Unit) -> Unit {
+    return interceptors.reversed().fold(initialInterceptor) { a, b ->
+      { context: CONTEXT, testCase: () -> Unit ->
+        b(context, { a.invoke(context, { testCase() }) })
+      }
+    }
+  }
+
+  // TODO beautify
+  fun runTest(spec: Spec, testCase: TestCaseDescriptor) {
+    if (testCase.isActive) {
+      val executor =
+          if (testCase.config.threads < 2) Executors.newSingleThreadExecutor()
+          else Executors.newFixedThreadPool(testCase.config.threads)
+      listener.executionStarted(testCase)
+      val initialInterceptor = { context: TestCaseContext, test: () -> Unit ->
+        spec.interceptTestCase(context, { test() })
+      }
+      val testInterceptorChain = createInterceptorChain(testCase.config.interceptors, initialInterceptor)
+      val testCaseContext = TestCaseContext(spec, testCase)
+
+      val errors = mutableListOf<Throwable>()
+
+      for (j in 1..testCase.config.invocations) {
+        executor.execute {
+          try {
+            testInterceptorChain(testCaseContext, { testCase.test() })
+          } catch (t: Throwable) {
+            errors.add(t)
```

Comment:
Shouldn't this be synchronized? Most of the time it will probably work with multiple threads, but there is no guarantee.